python_modules/dagster/dagster/_daemon/asset_daemon.py

```diff
@@ -993,61 +1013,128 @@ def _evaluate_auto_materialize_tick(
+        # check that the run_request requires the backfill daemon rather than if the setting is enabled to
+        # account for the setting changing between tick retries
+        if run_request.requires_backfill_daemon():
+            asset_graph_subset = check.not_none(run_request.asset_graph_subset)
+            if instance.get_backfill(reserved_run_id):
+                self._logger.warn(
+                    f"Run {reserved_run_id} already submitted on a previously interrupted tick, skipping"
+                )
             else:
-                submitted_run = submit_asset_run(
-                    run_id=reserved_run_id,
-                    run_request=run_request._replace(
+                instance.add_backfill(
+                    PartitionBackfill.from_asset_graph_subset(
+                        backfill_id=reserved_run_id,
+                        dynamic_partitions_store=instance,
+                        backfill_timestamp=get_current_timestamp(),
+                        asset_graph_subset=asset_graph_subset,
                         tags={
                             **run_request.tags,
                             AUTO_MATERIALIZE_TAG: "true",
                             AUTOMATION_CONDITION_TAG: "true",
                             ASSET_EVALUATION_ID_TAG: str(evaluation_id),
-                        }
-                    ),
-                    run_request_index=i,
-                    instance=instance,
-                    workspace_process_context=workspace_process_context,
-                    run_request_execution_data_cache=run_request_execution_data_cache,
-                    debug_crash_flags=debug_crash_flags,
-                    logger=self._logger,
+                        },
+                        title=f"Run for Declarative Automation evaluation ID {evaluation_id}",
+                        description=None,
+                    )
                 )
-                submitted_run_id = submitted_run.run_id
-                entity_keys = [
-                    *(run_request.asset_selection or []),
-                    *(run_request.asset_check_keys or []),
-                ]
-            # heartbeat after each submitted runs
+            return reserved_run_id, check.not_none(asset_graph_subset.asset_keys)
+        else:
+            submitted_run = submit_asset_run(
+                run_id=reserved_run_id,
+                run_request=run_request._replace(
+                    tags={
+                        **run_request.tags,
+                        AUTO_MATERIALIZE_TAG: "true",
+                        AUTOMATION_CONDITION_TAG: "true",
+                        ASSET_EVALUATION_ID_TAG: str(evaluation_id),
+                    }
+                ),
+                run_request_index=i,
+                instance=instance,
+                workspace_process_context=workspace_process_context,
+                run_request_execution_data_cache=run_request_execution_data_cache,
```

Comment:
is this cache thread safe?

Reply:
my understanding here is that the risk with this cache is simply that multiple threads will end up making redundant queries for execution data.

in theory, the values returned by these separate queries should be identical, and there's no read -> modify -> write sort of pattern happening here, which is where the non-thread-safety of dictionaries would come into play. Given that all writes into the same key of the dictionary should theoretically contain the same value, that shouldn't be an issue.